src/main/java/com/linecorp/armeria/common/ServiceInvocationContextProgressivePromise.java

```diff
@@ -0,0 +1,169 @@
+package com.linecorp.armeria.common;
+
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Stream;
+
+import io.netty.util.concurrent.Future;
+import io.netty.util.concurrent.GenericFutureListener;
+import io.netty.util.concurrent.ProgressivePromise;
+
+final class ServiceInvocationContextProgressivePromise<T> implements ProgressivePromise<T> {
+
+    private final ServiceInvocationContext context;
+    private final ProgressivePromise<T> delegate;
+
+    ServiceInvocationContextProgressivePromise(ServiceInvocationContext context,
+                                               ProgressivePromise<T> delegate) {
+        this.context = context;
+        this.delegate = delegate;
+    }
+
+    @Override
+    public ProgressivePromise<T> setProgress(long progress, long total) {
+        return delegate.setProgress(progress, total);
+    }
+
+    @Override
+    public boolean tryProgress(long progress, long total) {
+        return delegate.tryProgress(progress, total);
+    }
+
+    @Override
+    public ProgressivePromise<T> setSuccess(T result) {
+        return delegate.setSuccess(result);
+    }
+
+    @Override
+    public ProgressivePromise<T> setFailure(Throwable cause) {
+        return delegate.setFailure(cause);
+    }
+
+    @Override
+    public ProgressivePromise<T> addListener(
+            GenericFutureListener<? extends Future<? super T>> listener) {
+        return delegate.addListener(context.makeContextAware(listener));
```

Comment:
This will not work because we added a wrapped listener and a user will specify a unwrapped listener here. The default Promise implementation performs identity comparison, so overriding equals/hashCode() won't work as expected.